core/arch/arm/plat-zynqmp/drivers/aes.c

```diff
@@ -0,0 +1,180 @@
+#define CSU_AES_SIZE			0x2000
+
+/* CSU AES registers */
+#define AES_STS_OFFSET			0x00
+#define AES_KEY_SRC_OFFSET		0x04
+#define AES_KEY_LOAD_OFFSET		0x08
+#define AES_START_MSG_OFFSET		0x0C
+#define AES_RESET_OFFSET		0x10
+#define AES_KEY_CLR_OFFSET		0x14
+#define AES_CFG_OFFSET			0x18
+
+#define AES_STS_AES_BUSY	        BIT(0)
+#define AES_KEY_LOAD			BIT(0)
+#define AES_STS_AES_KEY_ZERO		BIT(8)
+#define AES_STS_KUP_ZEROED		BIT(9)
+#define AES_STS_KEY_INIT_DONE		BIT(4)
+#define AES_START_MSG			BIT(0)
+#define AES_CFG_ENC			BIT(0)
+#define AES_CFG_DEC			0
+#define AES_RESET_SET			BIT(0)
+#define AES_RESET_CLR			0
+
+/* CSU registers */
+#define SSS_OFFSET			0x008
+#define DMA_RESET_OFFSET		0x00c
+#define CSU_LOCAL_SIZE			0x20
+
+#define DMA0_STREAM_TO_AES		0x5A0
+#define DMA_RESET_SET			BIT(0)
+#define DMA_RESET_CLR			0
+
+#define AES_TIMEOUT_USEC		200000
+
+/* AES GCM */
+#define GCM_TAG_SIZE			16
+
+static TEE_Result aes_wait(uint32_t event, bool set)
+{
+	vaddr_t aes = core_mmu_get_va(CSU_AES_BASE, MEM_AREA_IO_SEC,
+				      CSU_AES_SIZE);
+	uint64_t tref = timeout_init_us(AES_TIMEOUT_USEC);
+	uint32_t status = 0;
+
+	while (!timeout_elapsed(tref)) {
+		status = io_read32(aes + AES_STS_OFFSET) & event;
+		if ((set && status == event) || (!set && status != event))
+			return TEE_SUCCESS;
+	}
+
+	return TEE_ERROR_GENERIC;
+}
+
+static TEE_Result aes_transfer(uint8_t *src, uint8_t *dst, uint8_t *iv,
+			       size_t len)
+{
+	TEE_Result ret = TEE_SUCCESS;
+
+	csu_dma_prepare();
+	csu_dma_transfer(CSU_DMA_SRC_CHANNEL, iv, GCM_TAG_SIZE / 4, 0);
+	ret = csu_dma_sync(CSU_DMA_SRC_CHANNEL);
+	if (ret) {
+		EMSG("dma iv transfer timeout");
+		return ret;
+	}
+	csu_dma_transfer(CSU_DMA_DST_CHANNEL, dst, (len + GCM_TAG_SIZE) / 4, 0);
+	csu_dma_transfer(CSU_DMA_SRC_CHANNEL, src, len / 4, CSU_DMA_DONE);
+	ret = csu_dma_sync(CSU_DMA_SRC_CHANNEL);
+	if (ret) {
+		EMSG("dma source transfer timeout");
+		return ret;
+	}
+	ret = csu_dma_sync(CSU_DMA_DST_CHANNEL);
+	if (ret) {
+		EMSG("dma destination transfer timeout");
+		return ret;
+	}
+	csu_dma_unprepare();
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result aes_encrypt_data(uint8_t *src, uint8_t *dst, uint8_t *iv,
+			    size_t src_len, size_t dst_len, size_t iv_len,
+			    uint32_t key)
+{
+	vaddr_t aes = core_mmu_get_va(CSU_AES_BASE, MEM_AREA_IO_SEC,
+				      CSU_AES_SIZE);
+	vaddr_t csu = core_mmu_get_va(CSU_BASE, MEM_AREA_IO_SEC,
+				      CSU_LOCAL_SIZE);
+	TEE_Result ret = TEE_SUCCESS;
+	uint32_t val = 0;
+
+	if (key != CSU_AES_KEY_SRC_DEV) {
+		EMSG("key type not supported");
+		return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+
+	if (src_len % 4) {
+		EMSG("invalid source size");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (iv_len < GCM_TAG_SIZE) {
+		EMSG("invalid iv size");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (dst_len < (src_len + GCM_TAG_SIZE)) {
+		EMSG("invalid destination size");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	/* connect DMA0 in/out to AES */
```

Comment:
Minor: I prefer to start comments and `EMSG()` text etc. with an uppercase letter